sips/drafts/sip-inlinemeta.md

```diff
@@ -0,0 +1,191 @@
+---
+layout: sip
+title: Inline Definitions and Meta Expressions
+---
+
+__Eugene Burmako__
+__Sébastien Doeraene__
+__Vojin Jovanovic__
+__Martin Odersky__
+__Dmitry Petrashko__
+__Denys Shabalin__
+
+__May 2016__
+
+# SIP NN: Inline Definitions and Meta Expressions (Public Draft)
+
+## Motivation ##
+
+## Inline Modifiers
+
+We introduce a new reserved word: `inline`. `inline` can be used as a modifier for:
+
+ - concrete value definitions, e.g.
+
+        inline val x = 4
+
+ - concrete methods, e.g.
+
+        inline def square(x: Double) = x * x
+
+ - parameters of inline methods, e.g.
+
+        inline def pow(b: Double, inline n: Int): Double = {
+          if (n == 0) 1
+          else pow(b, n - 1)
+        }
+
+The previous rule of regarding a `final val` with no explicit type as
+a compile-time constant (inherited from Java) is dropped. Instead we
+write such `val`s now as `inline`.
```

Comment:
Please note that another thing that `final val` does is make its type more precise:

```scala
scala> val x = 1
x: Int = 1

scala> final val x = 1
x: Int(1) = 1
```

So you would be throwing that out as well..